jep/0000/README.adoc

```diff
@@ -0,0 +1,243 @@
+with a huge number of interrelated components (mainly, though not exclusively, plugins) and many maintainers:
+
+* Every maintainer must obtain (and keep secure) personal credentials to Artifactory in order to perform releases.
+  Each maintainer also needs to be listed in `repository-permissions-updater`,
+  even if they have full rights to the GitHub repository.
+* A release involves running tests locally (e.g., on a laptop).
+  CPU-intensive test runs can interfere with other work;
+  flaky tests might behave differently than in on ci.jenkins.io.
+* It is impossible to verify that released binaries actually correspond to the purported tagged sources.
+* Developers need to pick version numbers whether or not the numbers have any meaning to humans.
+* Every release must have two machine-generated Git commits,
+  since the version number is encoded in the version-controlled `pom.xml`.
+* It is tricky to `git bisect` a problem reproduced only by a _dependency_ on a component
+  since even local builds (`mvn install`) get various version numbers.
+* The `perform` phase cannot be simulated in advance,
+  and involves procedures never used in other circumstances,
+  so often things go wrong and leave a “botched release” which must be cleaned up and retried.
+  Novice and even experienced Jenkins developers often have to ask for help with release problems.
+* Pull requests are sometimes approved and merged but https://github.com/jenkinsci/junit-attachments-plugin/pull/24#issuecomment-654900899[left unreleased for months on end],
+  merely because the PR’s value did not seem justified by the overhead of performing a release.
+
+The concept of continuous delivery (CD) is that “releases” should be cheap and frequent.
+Rather than forcing a maintainer to run a slow and perilous command every time anything of value is merged,
+harness automation to publish the newest source changes right away and without intervention.
+
+== Reasoning
+
+This JEP is inspired by link:../221/README.adoc[JEP-221],
+and shares similar motivation,
+but diverges from that proposal in various technical aspects.
+
+The use of JEP-305-based version numbers is attractive in that it requires no maintenance whatsoever:
+merely pushing a commit to `master` (including merging a pull request)
+suffices to trigger a deployment,
+and the version number will uniquely and securely identify that commit,
+with no need to create a redundant Git tag.
+Component maintainers who wish to follow SemVer principles,
+encoding some semantics into version numbers,
+can still do so by appending the generated number as a “micro” component to a manually maintained `major.minor.` prefix.
+
+GitHub Actions are attractive in this context because they define a trust boundary naturally scoped to the repository:
+a given bot token is defined in only repository, useful in only that repository, and used only for a containerized build of that repository.
```

Comment:
Such a detail can be abstracted through a secrets manager like Vault (or Azure Key Vault like in the release infra) for generating deploy keys for each run (as a neat side-effect, we could also sign the artifacts on demand using repo-specific certificates).

Note that this abstraction works for both the GitHub Actions workflow as well as a Jenkins workflow.

Reply:
What kind of “deploy keys” does Artifactory support beyond what is already outlined in this document? And what do you mean by “each run”? I do not think I am following your suggestion.

Reply:
As in a temporary deploy token for each run of the action.

Reply:
I do not see any practical way to accomplish that. The token would have to be generated on demand while the action runs by some privileged service which would have to establish the identity of the running action?

Reply:
My idea is something like, suppose you're building `jenkinsci/foo`. Then the action or pipeline will request a deploy key for that particular project which has limited authorization to upload files to expected locations as before. If the build attempted to deploy code elsewhere, it would find itself unauthorized at this point. If you're planning to create tokens for each repo, that gets you halfway there. The dynamic secrets binding would help with credentials rotation, dynamic creation of tokens, and better authorization integration.

Do note that this idea I'm proposing is still likely overkill. I'm considering different methods of using an external secrets manager in an unrelated task, and I thought I'd share the idea in a JEP that's exploring a similar space.

Reply:
> request a deploy key for that particular project

This is the problem. Request from some service TBD. And how does that service verify that the request is in fact coming from a GH Action run on that repository? https://github.community/t/verifying-connections-from-a-github-hosted-runner/17331 is the closest I can find, and basically unanswered. Can you send your `GITHUB_TOKEN` to that service and have it call back to the GH API and verify that the token has write access to the repo?

Reply:
It's quite possible that GH Actions are too new to have integrations with Vault or other things.

Reply:
Well I presume we would need to write any service ourselves anyway, since what it needs to do is generate Artifactory tokens with permissions driven by information in RPU. The question is whether there is some reliable way to have a GH Action run prove its authenticity to an external service.

If we find a way to do this, I do not dispute that it would be nicer than having the RPU build process save long-lived tokens to the repo.

Reply:
Yeah, there would certainly need to be some glue somewhere. See also the GH Action for Vault: https://github.com/marketplace/actions/vault-secrets

Reply:
I found that you can run

```bash
curl -s -H 'Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}' https://api.github.com/installation/repositories | jq -r '.repositories[0].full_name'
```

to prove that a supplied token is coming from an app installed on the repository. I am however having trouble verifying that the app has write permissions to the repository; https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#get-the-authenticated-app is not available using this installation token (it requires a JWT). https://docs.github.com/en/free-pro-team@latest/rest/reference/apps#get-an-app requires you to know the app id. I could try to call some other endpoint which requires write access to contents but this is kind of scary. The following seems to work:

```bash
TAG=permcheck-$RANDOM
curl -i -H 'Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}' -d '{"ref":"refs/tags/'$TAG'","sha":"'$GITHUB_SHA'"}' https://api.github.com/repos/$GITHUB_REPOSITORY/git/refs
curl -i -H 'Authorization: Bearer ${{ secrets.GITHUB_TOKEN }}' -X DELETE https://api.github.com/repos/$GITHUB_REPOSITORY/git/refs/tags/$TAG
```

I will propose this system in the document.

Reply:
> The dynamic secrets binding would help with credentials rotation, dynamic creation of tokens, and better authorization integration.

Isn't the credential rotation more of a problem in a large, sprawling, heterogeneous environment? We have a single GH org and a known subset of repos, all with API accessible at all times.

While this approach looks interesting (and I've also considered the need for rotating secrets), the only advantage I can see is this is more likely to not run into untimely secret rotation problems. Or perhaps more likely, if actions are allowed to run in parallel.

Reply:
Well my hope would be that the Artifactory token provisioning service would generate short-lived tokens, good for only one hour say, and each GH Action run would get a fresh token permitted only to deploy to the paths specified in RPU. I am pretty fuzzy on what is possible on the Artifactory side, though (and have no permissions to experiment).

Reply:
> each GH Action run would get a fresh token permitted only to deploy to the paths specified in RPU

That's the plan. There's some new stuff in Artifactory (time-limited access tokens) which would simplify this quite a bit—and could make a compelling case for a "credentials service". I'll need to check exactly what the requirements would be.

As a first step, we can store a credential in the GH secrets directly, and periodically update/rotate them from a locally executed script somewhere, without needing to operate a separate service.

Reply:
> from a locally executed script

If this is safer than the current proposal of doing this automatically from RPU builds, then sure.

Reply:
> So far they don't seem to be possible credentials for Maven, just the API, which is a weird design choice for Artifactory.

PEBCAK: I have two users, used one's username and the other's API token 🤦 

So API tokens work. Now on to using expiring access tokens…

Reply:
https://repo.jenkins-ci.org/webapp/#/artifacts/browse/tree/General/snapshots/io/jenkins/plugins/solarized-theme/0.2-SNAPSHOT/solarized-theme-0.2-20201021.162950-2.hpi is a snapshot of a plugin I uploaded using a token that expired after an hour (and the user name is something I used when generating the token, there's no actual token by that name).

The problem right now seems that it's not possible to assign permissions from a "permission target" to an access token, just a group membership.

RPU will need a larger overhaul for this, I think, between generating groups, poking GitHub's secrets API to update the deployment secret valid for the next X hours (RPU run interval + 3 hours or so?), and changing the YAML file format to accommodate multi-module projects and the new automated builds.

Reply:
> changing the YAML file format to accommodate multi-module projects

Actually necessary? There is already a `github` field which can be used to aggregate YAML files pertaining to the same repo.

Reply:
Since we are not leaning toward a pull-based approach, I am closing this thread and trying to capture outstanding design issues in the document.